### PR TITLE
Add complex64 and complex128 dtypes

### PR DESCRIPTION
**Pull request by ezyang**

### CODE REVIEW

File: aten/src/ATen/core/ScalarType.h [hunk 1]
```diff
@@ -6,20 +6,35 @@
...
 _(uint8_t,Byte,i)  /* 0 */ \
 _(int8_t,Char,i)   /* 1 */ \
 _(int16_t,Short,i) /* 2 */ \
 _(int,Int,i)       /* 3 */ \
 _(int64_t,Long,i)  /* 4 */ \
 _(at::Half,Half,d) /* 5 */ \
 _(float,Float,d)   /* 6 */ \
-_(double,Double,d) /* 7 */
+_(double,Double,d) /* 7 */ \
+_(std::complex<float>,ComplexFloat,z)   /* 8 */ \
```
Comment (SsnL) [inline aten/src/ATen/core/ScalarType.h]: No `ComplexHalf`? cufft supports that at least.

**Comment by SsnL** [comment]
Why is there a revert commit?

**Comment by ezyang** [comment]
Yeah, this can't merge as is, there's a temporary revert so that I can actually do local dev.

ComplexHalf seems like a reasonable thing to add for future compatibility.

**Comment by ezyang** [comment]
Hmm, ComplexHalf is a bit troublesome to use with `std::complex`, because it only has three well defined specializations and `at::Half` ain't one of them. @Roger-Luo, would you happen to know what a good C++ type to use in this case would be?

**Comment by SsnL** [comment]
One way would be do the same thing we did for `half` and create `at::Half2` (cuda uses `half2`).

**Comment by ezyang** [comment]
Yeah, that's what I ended up doing in this patch.

**Review by ssnl — APPROVED** [review]
LGTM

**Pull request closed without merging**